In [25]:
import pandas as pd
import numpy as np
df = pd.read_csv("Housing.csv")

In [26]:
df.head(10)

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,yes,no,no,no,yes,2,yes,furnished
1,12250000,8960,4,4,4,yes,no,no,no,yes,3,no,furnished
2,12250000,9960,3,2,2,yes,no,yes,no,no,2,yes,semi-furnished
3,12215000,7500,4,2,2,yes,no,yes,no,yes,3,yes,furnished
4,11410000,7420,4,1,2,yes,yes,yes,no,yes,2,no,furnished
5,10850000,7500,3,3,1,yes,no,yes,no,yes,2,yes,semi-furnished
6,10150000,8580,4,3,4,yes,no,no,no,yes,2,yes,semi-furnished
7,10150000,16200,5,3,2,yes,no,no,no,no,0,no,unfurnished
8,9870000,8100,4,1,2,yes,yes,yes,no,yes,2,yes,furnished
9,9800000,5750,3,2,4,yes,yes,no,no,yes,1,yes,unfurnished


In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


In [28]:
df.describe()

,price,area,bedrooms,bathrooms,stories,parking
count,5.450000e+02,545.000000,545.000000,545.000000,545.000000,545.000000
mean,4.766729e+06,5150.541284,2.965138,1.286239,1.805505,0.693578
std,1.870440e+06,2170.141023,0.738064,0.502470,0.867492,0.861586
min,1.750000e+06,1650.000000,1.000000,1.000000,1.000000,0.000000
25%,3.430000e+06,3600.000000,2.000000,1.000000,1.000000,0.000000
50%,4.340000e+06,4600.000000,3.000000,1.000000,2.000000,0.000000
75%,5.740000e+06,6360.000000,3.000000,2.000000,2.000000,1.000000
max,1.330000e+07,16200.000000,6.000000,4.000000,4.000000,3.000000


np.int8 – диапазон от -128 до 127;
np.int16 – диапазон от -32768 до 32767;
np.int32 – диапазон от -2.1·10⁹ до 2.1·10⁹;
np.int64 – диапазон от -9.2·10¹⁸ до 9.2·10¹⁸;
np.uint8 – диапазон от 0 до 255;
np.uint16 – диапазон от 0 до 65535;
np.uint32 – диапазон от 0 до 4.2·10⁹;
np.uint64 – диапазон от 0 до 1.8·10¹⁹;
1) Столбец price: min - 1.750000e+06; max - 1.330000e+07. Подходит диапазон от -2.1·10⁹ до 2.1·10⁹ (np.int32).
2) Столбец area: min - 1650.000000; max - 16200.000000. Подходит диапазон от -32768 до 32767 (np.int16).
3) Столбец bedrooms: min - 1.000000; max - 6.000000. Подходит диапазон от -128 до 127 (np.int8).
4) Столбец bathrooms: min - 1.000000; max - 4.000000. Подходит диапазон от -128 до 127 (np.int8).
5) Столбец stories: min - 1.000000; max - 4.000000. Подходит диапазон от -128 до 127 (np.int8).
6) Столбец parking: min - 0.000000; max - 3.000000. Подходит диапазон от -128 до 127 (np.int8).

7) Для оптимизации столбца furnishingstatus можно использовать переход к категориальным значениям – так как в столбце используются повторяющиеся однотипные элементы (furnished, unfurnished, semi-furnished).

8) В остальных столбцах существуют только значения 'yes' и 'no', поэтому можно использовать переход к булевым значениям.

In [29]:
df['price'] = df['price'].astype(np.int32)
df['area'] = df['area'].astype(np.int16)
df['bedrooms'] = df['bedrooms'].astype(np.int8)
df['bathrooms'] = df['bathrooms'].astype(np.int8)
df['stories'] = df['stories'].astype(np.int8)
df['parking'] = df['parking'].astype(np.int8)
df['furnishingstatus'] = df['furnishingstatus'].astype('category')
categories = ['mainroad', 'guestroom', 'basement', 'hotwaterheating', 'airconditioning', 'prefarea']
for x in categories:
    df[x] = df[x].map({'yes': True, 'no': False}).astype('bool')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   price             545 non-null    int32   
 1   area              545 non-null    int16   
 2   bedrooms          545 non-null    int8    
 3   bathrooms         545 non-null    int8    
 4   stories           545 non-null    int8    
 5   mainroad          545 non-null    bool    
 6   guestroom         545 non-null    bool    
 7   basement          545 non-null    bool    
 8   hotwaterheating   545 non-null    bool    
 9   airconditioning   545 non-null    bool    
 10  parking           545 non-null    int8    
 11  prefarea          545 non-null    bool    
 12  furnishingstatus  545 non-null    category
dtypes: bool(6), category(1), int16(1), int32(1), int8(4)
memory usage: 9.3 KB


До оптимизации потребление памяти составляло 55.5+ KB, а после 9.3 KB. То есть потребление памяти уменьшилось примерно на 46 KB.